feat: Destination tests for batch exports

tomasfarias opened the pull request:
## Problem

The most common type error when interacting with batch exports are user configuration errors. This PR introduces a new framework for "Destination tests": Tests that can be called from our API to evaluate a given batch export configuration.

<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

* [x] Bigquery destination tests
* [x] Snowflake destination tests
* [x] S3 destination tests
* [ ] PostgreSQL destination tests
* [ ] Redshift destination tests
* [x] Frontend support
* [x] API Support

Frontend screenshots:

All passed:
![image](https://github.com/user-attachments/assets/989d7a8d-3eb9-4069-a153-f192427e0024)

Error:
![image](https://github.com/user-attachments/assets/11bb662b-617e-4b9c-b3fc-269614618190)


<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?

<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

New unit tests.
<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
refactor(bigquery): Move some of client initialization to client class

tomasfarias committed:
feat: Backend for BigQuery destination tests

tomasfarias committed:
fix: Re-add a bunch of master changes

github-actions[bot] committed:
Update query snapshots

tomasfarias committed:
feat: S3 destination tests

rossgray commented on posthog/batch_exports/http.py:
```diff
@@ -526,6 +526,36 @@ def perform_destroy(self, instance: BatchExport):
         """
         disable_and_delete_export(instance)
 
+    @action(methods=["POST"], detail=True)
+    def test(self, request: request.Request, *args, **kwargs) -> response.Response:
```
is this endpoint just for fetching a test (and it's list of steps) for a given batch export?

tomasfarias replied:
Yeah, thought about having everything in a single endpoint, but it looked too complicated. I'd be happy to rename this to something more verbose though...

rossgray replied:
I think it does make sense to get a list of possible tests upfront, especially if we want to add them to destinations one at a time, since the frontend could probably not even show a test button if there are no tests to run for example. Was also wondering if it could just be a GET method if it's not actually running anything

tomasfarias committed:
test: API tests for destination tests

tomasfarias committed:
refactor: Make configure a required method, change signature

tomasfarias committed:
fix: Configure test before running step

github-actions[bot] committed:
Update query snapshots

tomasfarias committed:
feat: Add frontend for destination tests

This adds a test configuration component in the batch exports
configuration page. Currently, tests are optional.

tomasfarias committed:
fix: Don't display anything if there are no destination tests

tomasfarias committed:
fix: Re-run all steps from the beginning

Eventually, we could allow partial retries

tomasfarias committed:
fix: Return NOT_FOUND for destinations without tests

tomasfarias committed:
fix: Test now expects step instead of result

tomasfarias committed:
fix: Typing errors in frontend

tomasfarias committed:
fix: More typing errors in frontend

tomasfarias committed:
chore: Add a bunch of docstrings

rossgray commented on posthog/temporal/batch_exports/destination_tests.py:
```diff
@@ -0,0 +1,466 @@
+
+    async def run(self) -> DestinationTestStepResult:
+        """Run this test step."""
+        import aioboto3
```
Same question about imports here

tomasfarias replied:
In this particular case, I liked the fact that the imports are included as part of the test itself.

rossgray replied:
for organizational reasons?

tomasfarias replied:
In case we ever introduce errors that are raised on import, I rather that cause a test failure just for the relevant test.

rossgray replied:
I guess that's a valid point, although ideally we catch these in our unit tests

rossgray reviewed (commented):
Looks good so far. Just left some minor comments but like the structure of the code a lot.

My only other comment would be that it's perhaps a bit confusing that we refer to a `DestinationTest` (singular) as a set of different tests for a destination but again not a deal breaker.

tomasfarias commented on posthog/batch_exports/http.py:
```diff
@@ -526,6 +526,38 @@ def perform_destroy(self, instance: BatchExport):
+        test_step = request.data.pop("step", 0)
+
+        serializer = self.get_serializer(data=request.data)
+        _ = serializer.is_valid(raise_exception=True)
```
The tradeoff is in being explicit. My IDE was complaining about implicitly not using the return value here.

rossgray replied:
hmm ok, fair enough

tomasfarias committed:
refactor: Make name and description class variables

tomasfarias committed:
refactor: Move imports to the top

tomasfarias commented on posthog/batch_exports/http.py:
```diff
@@ -526,6 +526,38 @@ def perform_destroy(self, instance: BatchExport):
 
+    @action(methods=["POST"], detail=False, required_scopes=["INTERNAL"])
+    def test(self, request: request.Request, *args, **kwargs) -> response.Response:
+        from posthog.temporal.batch_exports.destination_tests import get_destination_test
```
In this case I don't see a reason (yes in the one below). Moved the imports to the top.

tomasfarias commented on posthog/batch_exports/http.py:
```diff
@@ -526,6 +526,38 @@ def perform_destroy(self, instance: BatchExport):
         """
         disable_and_delete_export(instance)
 
+    @action(methods=["POST"], detail=False, required_scopes=["INTERNAL"])
```
It could, yeah, but it would require creating one endpoint per destination, at least with how things are setup right now. Given that it's an internal API, I'd just go with this for simplicity.

rossgray replied:
Why's that? Could you not accept a query parameter?

tomasfarias commented on posthog/temporal/batch_exports/destination_tests.py:
```diff
@@ -0,0 +1,466 @@
+        aws_secret_access_key: str | None = None,
+    ) -> None:
+        super().__init__(
+            name="Check S3 bucket exists", description="Verify the configured S3 bucket exists and we can access it"
```
I agree with this. Also expanded the descriptions a bit.

tomasfarias committed:
refactor: Separate branching logic from step execution

tomasfarias committed:
feat: Add Snowflake destination tests

tomasfarias committed:
fix: Type fixes

tomasfarias committed:
refactor: Switch to GET for test endpoint

github-actions[bot] committed:
Update query snapshots

rossgray reviewed (approved):
Looks good, let's ship it and reduce the number of errors we get!
